nvtabular/workflow/node.py

```diff
@@ -65,16 +67,105 @@ def selector(self, sel):
+                    elif isinstance(nested_dep, WorkflowNode):
+                        schema += nested_dep.output_schema
+        return schema
+
+    @property
+    def parents_schema(self):
+        return sum([parent.output_schema for parent in self.parents], Schema())
+
     def compute_schemas(self, root_schema):
+        # If parent is an addition node, we may need to propagate grouping
+        # unless we're a node that already has a selector
+        if not self.selector:
+            if (
+                len(self.parents) == 1
+                and isinstance(self.parents[0].op, internal.ConcatColumns)
+                and self.parents[0].selector
+                and self.parents[0].selector.names
+            ):
+
+                self.selector = self.parents[0].selector
+
+        # If we have a selector, apply it to upstream schemas from nodes/dataset
         if self.selector:
             upstream_schema = Schema()
             upstream_schema += root_schema
             upstream_schema += self.parents_schema
-
+            upstream_schema += self.dependency_schema
             self.input_schema = upstream_schema.apply(self.selector)
         else:
-            self.input_schema = self.parents_schema
+            # If we don't have a selector but we're an addition node,
+            if isinstance(self.op, ConcatColumns):
+                upstream_selector = ColumnSelector()
+
+                for parent in self.parents:
+                    upstream_selector += ColumnSelector(parent.output_schema.column_names)
+
+                for dep in self.dependencies:
+                    if isinstance(dep, WorkflowNode):
+                        upstream_selector += ColumnSelector(dep.output_schema.column_names)
+                    elif isinstance(dep, ColumnSelector):
+                        upstream_selector += dep
+                    elif isinstance(dep, list):
+                        subgroup_selector = ColumnSelector()
+                        for elem in dep:
+                            if isinstance(elem, WorkflowNode):
+                                subgroup_selector += ColumnSelector(elem.output_schema.column_names)
+                            elif isinstance(elem, ColumnSelector):
+                                subgroup_selector += elem
+
+                        upstream_selector += ColumnSelector(subgroups=[subgroup_selector])
+
+                if upstream_selector.names:
+                    self.selector = upstream_selector
+
+                # For addition nodes, some of the operands are parents and
+                # others are dependencies so grab schemas from both
+                self.input_schema = self.parents_schema + self.dependency_schema
+
+            # If we're a subtraction node, we have to do some gymnastics to compute
+            # the schema, because operands may be in the parents or the dependencies
+            # or both
+            elif isinstance(self.op, SubsetColumns):
```

Comment:
These chunks that compute the schemas should be things we can do on the `Operator` sub-classes themselves